src/iface/route.rs

```diff
@@ -0,0 +1,207 @@
+/// use smoltcp::iface::Routes;
+/// let mut routes = Routes::new(BTreeMap::new());
+/// ```
+///
+/// On systems without heap, use:
+///
+/// ```rust
+/// use smoltcp::iface::Routes;
+/// let mut routes_storage = [];
+/// let mut routes = Routes::new(&mut routes_storage[..]);
+/// ```
+#[derive(Debug)]
+pub struct Routes<'a> {
+    storage:      ManagedMap<'a, IpCidr, Route>,
+}
+
+impl<'a> Routes<'a> {
+    /// Creates a routing tables. The backing storage is **not** cleared
+    /// upon creation.
+    pub fn new<T>(storage: T) -> Routes<'a>
+            where T: Into<ManagedMap<'a, IpCidr, Route>> {
+        let storage = storage.into();
+        Routes { storage }
+    }
+
+    /// Update the routes of this node.
+    ///
+    /// # Panics
+    /// This function panics if any of the addresses are not unicast.
+    pub fn update_routes<F: FnOnce(&mut ManagedMap<'a, IpCidr, Route>)>(&mut self, f: F) {
+        f(&mut self.storage);
+    }
+
+    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    ///
+    /// If not enough storage, returns the gateway address.
+    #[cfg(feature = "proto-ipv4")]
+    pub fn add_default_ipv4_gateway(&mut self, gateway: Ipv4Address) -> Result<Option<Route>, IpAddress> {
+        let cidr = IpCidr::new(IpAddress::v4(0, 0, 0, 0), 0);
+        let route = Route::new_ipv4_gateway(gateway);
+        match self.storage.insert(cidr, route) {
+            Ok(route) => Ok(route),
+            Err((_cidr, route)) => Err(route.via_router)
+        }
+    }
+
+    /// Add a default ipv6 gateway (ie. "ip -6 route add ::/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    ///
+    /// If not enough storage, returns the gateway address.
+    #[cfg(feature = "proto-ipv6")]
+    pub fn add_default_ipv6_gateway(&mut self, gateway: Ipv6Address) -> Result<Option<Route>, IpAddress> {
+        let cidr = IpCidr::new(IpAddress::v6(0, 0, 0, 0, 0, 0, 0, 0), 0);
+        let route = Route::new_ipv6_gateway(gateway);
+        match self.storage.insert(cidr, route) {
+            Ok(route) => Ok(route),
+            Err((_cidr, route)) => Err(route.via_router)
+        }
+    }
+
+    pub(crate) fn lookup(&self, addr: &IpAddress, timestamp: Instant) ->
+            Option<IpAddress> {
+        if addr.is_broadcast() {
+            panic!("IP address {} is broadcast", addr);
+        }
+
+        let cidr = match addr {
+            #[cfg(feature = "proto-ipv4")]
+            IpAddress::Ipv4(addr) => IpCidr::Ipv4(Ipv4Cidr::new(*addr, 32)),
+            #[cfg(feature = "proto-ipv6")]
+            IpAddress::Ipv6(addr) => IpCidr::Ipv6(Ipv6Cidr::new(*addr, 128)),
+            _ => unimplemented!()
+        };
+
+        for (prefix, route) in self.storage.range((Bound::Unbounded, Bound::Included(cidr))).rev() {
+            // TODO: do something with route.preferred_until
```

Comment:
You might want to create a FIXME or issue to research a clever way to clear out expired routes.

Reply:
What do you mean by "create a FIXME"?

Reply:
Add a comment with something like the following.

```
FIXME(ProgVal): Find a clever way to clear out expired routes.
```

I think an issue would be best though.

Reply:
Done: https://github.com/m-labs/smoltcp/issues/221